### PR TITLE
[libc++] Split the monolithic __threading_support header

### PR DESCRIPTION
**Pull request by ldionne**
The <__threading_support> header is a huge beast and it's really difficult to navigate. I find myself struggling to find what I want every time I have to open it, and I've been considering splitting it up for years for that reason.

This patch aims not to contain any functional change. The various implementations of the threading base are simply moved to separate headers and then the individual headers are simplified in mechanical ways. For example, we used to have redundant declarations of all the functions at the top of `__threading_support`, and those are removed since they are not needed anymore. The various #ifdefs are also simplified and removed when they become unnecessary.

Finally, this patch adds documentation for the API we expect from any threading implementation.

### CODE REVIEW

File: libcxx/include/__thread/support/c11.h [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+#include <__chrono/duration.h>
+#include <__config>
+#include <ctime>
+#include <threads.h>
+
+#ifndef _LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER
+#  pragma GCC system_header
+#endif
+
+typedef ::timespec __libcpp_timespec_t;
```
Comment (philnik777) [inline libcxx/include/__thread/support/c11.h:23]: Maybe refactor these to `using` aliases in a follow-up patch?

File: libcxx/include/__thread/support.h [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+int __libcpp_tls_create(__libcpp_tls_key*, void (*__at_exit)(void*));
+void* __libcpp_tls_get(__libcpp_tls_key);
+int __libcpp_tls_set(__libcpp_tls_key, void*);
+
+*/
+
+#if !defined(_LIBCPP_HAS_NO_THREADS)
+
+#  if defined(_LIBCPP_HAS_THREAD_API_EXTERNAL)
+#    include <__thread/support/external.h>
```
Comment (philnik777) [inline libcxx/include/__thread/support.h:106]: Maybe just include the `<__external_threading>` header directly? There doesn't seem to be much reason not to.

**Review by philnik777 — APPROVED** [review]
Thanks! This looks really nice. LGTM % nit.

**Comment by nico** [comment]
This requires updating libcxx and libcxxabi at the same time.

Any opposition to replacing the includes in libcxxabi with

```
#if __has_include(<__thread/support.h>)
#include <__thread/support.h>
#else
#include <__threading_support>
#endif
```

for a while, so that projects can keep rolling libcxxabi and libcxx independently? (We have an autoroller for libcxxabi that has been working without issues for years up to this change, and we have one for libcxx that's broken pretty frequently due to libcxx changes, including right now. It'd be nice if libcxxabi could keep rolling.)

**Comment by ldionne** [comment]
> This requires updating libcxx and libcxxabi at the same time.
> 
> Any opposition to replacing the includes in libcxxabi with
> 
> ```
> #if __has_include(<__thread/support.h>)
> #include <__thread/support.h>
> #else
> #include <__threading_support>
> #endif
> ```
> 
> for a while, so that projects can keep rolling libcxxabi and libcxx independently? (We have an autoroller for libcxxabi that has been working without issues for years up to this change, and we have one for libcxx that's broken pretty frequently due to libcxx changes, including right now. It'd be nice if libcxxabi could keep rolling.)

We can do that... but what I don't understand is how you build libc++abi. We require building libc++ and libc++abi at the same time, so how do you manage to roll both libraries independently?

**Comment by nico** [comment]
We don't use upstream's cmake files. We do build libc++ and libc++abi at the same time, but sometimes the revisions are off by a bit, since both are autoupdated independently.

Happy to go into as much detail as you want :)

**Comment by ldionne** [comment]
Would there be any appetite on your end to update both at the same time?

I'm definitely willing to add the workaround temporarily to ease your internal process since the cost is low, however that's not something we'd want to do on a regular basis going forward. In fact, I want to write a RFC for merging the libc++ and libc++abi projects, which would require this work on your end regardless.

**Comment by ldionne** [comment]
See 372f7dd48f0165c6a7cdc819516132c617ef6f45.

I will remove the work-around by EOW -- is that fine? I would really appreciate if you folks could take this opportunity to ensure that libc++ and libc++abi are rolled as a unit, since your setup is not supported right now. We really value the project's relationship with Chrome so it makes sense to do a temporary workaround like this, but I'm sure you understand how it's not viable to accommodate similar requests coming from arbitrary downstreams.

**Comment by nico** [comment]
> Would there be any appetite on your end to update both at the same time?

Unless absolutely necessary, no :) Updating LLVM can be a lot of work. The more parts of it we can ship independently, the better. libcxx is one of the harder parts to update, so I wouldn't want us to tie anything to it unless we want to.

(We'd prefer to move in the opposite direction: We currently bundle the sanitizer runtimes with the compiler/linker package, and it'd be nice if they'd update separately for example. The sanitizer runtimes have been pretty stable the last few years though, so it's not a priority.)

> I will remove the work-around by EOW -- is that fine?

Is it expensive to keep around for a bit longer? At the moment, updating libc++ is blocked on removing an is_pod somewhere in our codebase, but it's in a dependency that needs to roll in both my cherry-pick and a bunch of related things, and the new version of the dependency has some msan issues at the moment, etc. Fixing things in dependencies can take a while since we don't fully control them. (For example, https://crbug.com/1481891 – the `_LIBCPP_CHAR_TRAITS_REMOVE_BASE_SPECIALIZATION` thing – took three months to resolve)

**Comment by nico** [comment]
By coincidence, our libc++ roll blocker got resolved yesterday and we could roll in the 42 revisions that landed Jan 19 - 21 yesterday. And there were no new issues in the 49 revisions that landed Jan 23 - Feb 2, so we're now all caught up again and don't need this workaround any more. So end of week is actually good enough (barring reverts, but looks ok so far) – but it's hard to say in advance with changes in deps.

Anyways, fine to undo 372f7dd48f0165c6a7cdc819516132c617ef6f45 from our end. I don't know if it's useful for anyone else out there.

**Comment by nico** [comment]
(Workaround got reverted here: https://github.com/llvm/llvm-project/commit/c5f68a711c62aa1748c03215d95ad9b8c7dff9dd)

**Pull request merged**